src/server/req_delete.c

```diff
@@ -362,230 +390,288 @@ req_deletejob(struct batch_request *preq)
+			if (rc == TRUE)
+				continue;
+		}
 
-		/* single subjob, if running do full delete, */
-		/* if not then just set it expired		 */
 
-		offset = subjob_index_to_offset(parent, get_index_from_jid(jid));
-		if (offset == -1) {
-			req_reject(PBSE_UNKJOBID, 0, preq);
-			return;
-		}
-		subjobstate = get_subjob_state(parent, offset);
-		if (subjobstate == -1) {
-			req_reject(PBSE_IVALREQ, 0, preq);
-			return;
-		}
+		if (jt == IS_ARRAY_NO) {
 
-		if ((subjobstate == JOB_STATE_LTR_EXITING) && (forcedel == 0)) {
-			if (parent->ji_pmt_preq != NULL) {
-				pjob = find_job(jid);
-				reply_preempt_jobs_request(PBSE_BADSTATE, PREEMPT_METHOD_DELETE, pjob);
-			}
-			req_reject(PBSE_BADSTATE, 0, preq);
-			return;
-		} else if (subjobstate == JOB_STATE_LTR_EXPIRED) {
-			req_reject(PBSE_NOHISTARRAYSUBJOB, 0, preq);
-			return;
-		} else if ((pjob = parent->ji_ajtrk->tkm_tbl[offset].trk_psubjob)) {
-			/*
-			 * If the request is to also purge the history of the sub job then set ji_deletehistory to 1
+			/* just a regular job, pass it on down the line and be done
+			 * If the request is to purge the history of the job then set ji_deletehistory to 1
 			 */
 			if (delhist)
-				pjob->ji_deletehistory = 1;
-			req_deletejob2(preq, pjob);
-		} else {
-			acct_del_write(jid, parent, preq, 0);
-			parent->ji_ajtrk->tkm_tbl[offset].trk_substate = JOB_SUBSTATE_TERMINATED;
-			set_subjob_tblstate(parent, offset, JOB_STATE_LTR_EXPIRED);
-			parent->ji_ajtrk->tkm_dsubjsct++;
+				parent->ji_deletehistory = 1;
+			req_deletejob2(preq, parent);
+			continue;
 
-			decr_single_subjob_usage(parent);
+		} else if (jt == IS_ARRAY_Single) {
+			char subjobstate;
 
-			reply_ack(preq);
-		}
-		chk_array_doneness(parent);
-		return;
+			/* single subjob, if running do full delete, */
+			/* if not then just set it expired		 */
 
-	} else if (jt == IS_ARRAY_ArrayJob) {
-		/*
-		 * For array jobs the history is stored at the parent array level and also at the subjob level .
-		 * If the request is to delete the history of an array job then set  ji_deletehistory to 1 for
-		 * the parent array.The function chk_array_doneness() will take care of eventually
-		 *  purging the history .
-		 */
-		if (delhist)
-			parent->ji_deletehistory = 1;
-		/* The Array Job itself ... */
-		/* for each subjob that is running, delete it via req_deletejob2 */
+			offset = subjob_index_to_offset(parent, get_index_from_jid(jid));
+			if (offset == -1) {
+				update_deletejob_stat(jid, preq, PBSE_UNKJOBID, &preply->brp_un.brp_delstat);
+				if (preq->rq_ind.rq_delete.tot_rpys == preq->rq_ind.rq_delete.tot_jobs) {
+					req_reject(PBSE_UNKJOBID, 0, preq);  
+					return;
+				} else
+					continue;
+			}
 
-		++preq->rq_refct;
+			subjobstate = get_subjob_state(parent, offset);
+			if (subjobstate == -1) {
+				update_deletejob_stat(jid, preq, PBSE_IVALREQ, &preply->brp_un.brp_delstat);
+				if (preq->rq_ind.rq_delete.tot_rpys == preq->rq_ind.rq_delete.tot_jobs) {
```

Comment:
Looks like we are doing this if condition after every call to update_deletejob_stat(), can't move this in update_deletejob_stat() itself?

Reply:
Please refer to the code following this condition, in case of success, we need to reply to continue to the next jobid or we need to return the call for req_delete, still, we need to keep {if, else} block, that we can't avoid.